### PR TITLE
Registration and Login using JWT

### PR DESCRIPTION
**Pull request by abduljabbar-invozone**
Registration and login module is done. Implemented local auth guard passport strategy and getting access_token after verifying user model

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -24,9 +24,21 @@
     },
     "dependencies": {
         "@nestjs/common": "^8.0.0",
+        "@nestjs/config": "^1.0.1",
         "@nestjs/core": "^8.0.0",
+        "@nestjs/jwt": "^8.0.0",
+        "@nestjs/passport": "^8.0.0",
         "@nestjs/platform-express": "^8.0.0",
         "@nestjs/typeorm": "^8.0.1",
+        "@types/bcrypt": "^5.0.0",
```
Comment (isheraz) [inline package.json]: this will go into dev dependency

File: package.json [hunk 2]
```diff
@@ -24,9 +24,21 @@
...
+        "bcrypt": "^5.0.1",
+        "class-transformer": "^0.4.0",
+        "class-validator": "^0.13.1",
+        "cookie-parser": "^1.4.5",
```
Comment (isheraz) [inline package.json:36]: remove additional packages.

File: package.json [hunk 3]
```diff
@@ -38,9 +50,12 @@
         "@nestjs/cli": "^8.0.0",
         "@nestjs/schematics": "^8.0.0",
         "@nestjs/testing": "^8.0.0",
+        "@types/cookie-parser": "^1.4.2",
         "@types/express": "^4.17.13",
         "@types/jest": "^26.0.24",
         "@types/node": "^16.0.0",
+        "@types/passport-jwt": "^3.0.6",
```
Comment (isheraz) [inline package.json:57]: move all @types to devDep

File: src/app.controller.ts [hunk 4]
```diff
@@ -1,5 +1,6 @@
-import { Controller, Get } from '@nestjs/common';
+import { Controller, Get, Post, Request, UseGuards } from '@nestjs/common';
```
Comment (isheraz) [inline src/app.controller.ts]: remove if not required.

File: src/auth/local.strategy.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        super({
+            usernameField: 'email'
+        }); // config
+    }
+
+    async validate(username: string, password: string): Promise<any> {
+        const user = await this.authService.validateUser(username, password);
+
+        if(!user){
+            throw new UnauthorizedException();
```
Comment (isheraz) [inline src/auth/local.strategy.ts]: return throw new UnauthorizedException();

File: src/user/user.service.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+@Injectable()
+export class UserService {
+    constructor(
+        @InjectRepository(User)
+        private userRepository: Repository<User>,
+      ) {}
+
+      async findOne(email: string): Promise<User | undefined> {
+        return await this.userRepository.findOne({where: {email: email}});
```
Comment (isheraz) [inline src/user/user.service.ts]: make sure to throw appropriate error and return response in that error

**Review by isheraz — CHANGES_REQUESTED** [review]
Fix all comments and squash and rebase your commits. 
`rebase-squash`
1- squash your commits. `git rebase -i HEAD~<number_of_commits_to_squash>`
2- push the squashed commits (force push)
3- git rebase <latest-branch-maintined> (compare and continue)
4- git push and make a PR.

File: package.json (hunk 2 above)
Comment (abduljabbar-invozone) [reply package.json:36]: We need @nestjs/passport and passport both and apart from that we also need based on the strategy that we need to implement
passportjs

**Pull request merged**